homeassistant/components/mystrom/__init__.py

```diff
@@ -1 +1,82 @@
-"""The mystrom component."""
+"""The myStrom integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+import aiohttp
+from pymystrom.bulb import MyStromBulb
+from pymystrom.exceptions import MyStromConnectionError
+from pymystrom.switch import MyStromSwitch
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import PlatformNotReady
+from homeassistant.helpers.typing import ConfigType
+
+from .const import DOMAIN
+
+PLATFORMS: list[Platform] = [Platform.SWITCH]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def get_device_info(host: str) -> dict[str, Any]:
+    """Get the device info of a myStrom device."""
+    async with aiohttp.ClientSession() as session:
+        async with session.get(f"http://{ host}/api/v1/info") as response:
+
+            if response.status != 200:
+                raise MyStromConnectionError()
+            content_type = response.headers.get("Content-Type", "")
+            if "application/json" in content_type:
+                return await response.json()
+            raise MyStromConnectionError()
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the myStrom component."""
+    hass.data[DOMAIN] = {}
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up myStrom from a config entry."""
+    host = entry.data[CONF_HOST]
+    device = None
+    try:
+        info = await get_device_info(host)
+        device_type = info["type"]
+        if device_type in [101, 106, 107]:
+            device = MyStromSwitch(host)
+            await device.get_state()
+        elif device_type == 102:
+            mac = info["mac"]
+            device = MyStromBulb(host, mac)
+            await device.get_state()
+            if device.bulb_type not in ["rgblamp", "strip"]:
+                _LOGGER.error(
+                    "Device %s (%s) is not a myStrom bulb nor myStrom LED Strip",
+                    host,
+                    mac,
+                )
+                device = None
```

Comment:
Please move code out of the try block that isn't expected to raise

Reply:
I would appreciate a little hint on how to improve this one: `get_device_info` and the two `device.get_state` can both raise the `MyStromConnectionError`. I could wrap the 3 calls with a try block but this would look more complex to me. How can I best move code out of the try block without making the code more complex?

Reply:
It feels to me like maybe it should simply be:
```python
    try:
        info = await get_device_info(host)
    except MyStromConnectionError as err:
        _LOGGER.error("No route to myStrom plug: %s", host)
        raise ConfigEntryNotReady() from err

    device_type = info["type"]
    if device_type in [101, 106, 107]:
        device = MyStromSwitch(host)
    elif device_type == 102:
        mac = info["mac"]
        device = MyStromBulb(host, mac)
    else:
        # not sure what can be done here... maybe log error and return false?
        return False

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
        "device": device,
        "info": info,
    }
    hass.config_entries.async_setup_platforms(entry, PLATFORMS)

    return True
```

All the other checks should probably be inside the individual platforms.

Reply:
If you really feel that `device.get_state()` is required, then you need wrap just that call in a second and third `try...except`.

Reply:
Implemented as proposed